lib/common/xml.c

```diff
@@ -460,14 +460,91 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
     return NULL;
 }
 
+/*!
+ * \internal
+ * \brief Set an XML attribute, expanding \c ++ and \c += where appropriate
+ *
+ * If \p target already has an attribute named \p name set to an integer value
+ * and \p value is an addition assignment expression on \p name, then expand
+ * \p value to an integer and set attribute \p name to the expanded value in
+ * \p target.
+ *
+ * Otherwise, set attribute \c name on \p target using the literal \p value.
+ *
+ * For example, suppose \p target has an attribute named \c "X" with value
+ * \c "5", and that \p name is \c "X".
+ * * If \p value is \c "X++", the new value of \c "X" in \p target is \c "6".
+ * * If \p value is \c "X+=3", the new value of \c "X" in \p target is \c "8".
+ * * If \p value is \c "val", the new value of \c "X" in \p target is \c "val".
+ * * If \p value is \c "Y++", the new value of \c "X" in \p target is \c "Y++".
+ *
+ * \param[in,out] target  XML node whose attribute to set
+ * \param[in]     name    Name of the attribute to set
+ * \param[in]     value   New value of attribute to set
+ *
+ * \note The original attribute value in \p target and the number in an
+ *       assignment expression in \p value are parsed as scores. For more
+ *       details, refer to \c char2score().
+ */
+#if defined(PCMK__UNIT_TESTING)
+void
+#else
+static void
+#endif  // defined(PCMK__UNIT_TESTING)
+pcmk__xa_set_expand(xmlNode *target, const char *name, const char *value)
+{
+    // @TODO Consider moving to nvpair.c for logical separation
+    const char *old_value = NULL;
+
+    CRM_CHECK((target != NULL) && (name != NULL) && (value != NULL), return);
+
+    old_value = crm_element_value(target, name);
+
+    /* If no previous value, skip to default case and set the value unexpanded.
+     *
+     * @COMPAT Should we try to expand, assuming 0 for the old value? This
+     * doesn't seem fundamentally different from the "expanding ourselves" case
+     * mentioned below -- it's old=<unset>/new="X++" versus old="X++"/new="X++".
+     */
+    if (old_value != NULL) {
+        const char *n = name;
+        const char *v = value;
+
+        // Stop at first character that differs between name and value
+        for (; *n == *v; n++, v++);
+
+        // If value begins with name followed by a "++" or "+="
+        if ((*n == '\0')
+            && (*v++ == '+')
+            && ((*v == '+') || (*v == '='))) {
+
+            // If we're expanding ourselves, no previous value was set; use 0
+            int old_value_i = (old_value != value)? char2score(old_value) : 0;
+
+            /* value="X++": new value of X is old_value + 1
+             * value="X+=Y": new value of X is old_value + Y (for some number Y)
+             */
+            int add = (*v == '+')? 1 : char2score(++v);
```

Comment:
> which also means that updating "foo" to "foo+=bar" will result in "0"

It would result in `foo = foo`, ultimately. This is just determining what we add to `foo`

Reply:
if `foo` originally had a score in it; maybe you're talking about if it doesn't

Reply:
Right, if an attribute "foo" is initially set to the literal value "foo", `char2score()` will treat that as 0, and if it's set to the literal value "foo+=bar", `char2score()` will treat "bar" as 0, and the attribute will end up being set to 0